### PR TITLE
feat(dotfiles): run history reload commands after mise dot apply

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/mise/trails/60

`[history.reload]` maps a path glob to a command that runs once after files matching the glob change. It already ran after `mise dot pull`, `rollback`, `undo` and the watcher's automatic apply, but not after `mise dot apply`. A post-apply fixup such as `chmod 700 ~/.ssh` or an application reload therefore only happened when the files arrived through history, and a plain apply of a changed dotfile left the running application on its old configuration.

`mise dot apply` now runs the matching reload commands for the targets it actually wrote (symlinks created, copies written, templates rendered, edits applied), each once, after all writes.

```toml
[dotfiles]
"~/.config/hypr" = "dotfiles/hypr"
"~/.ssh/config" = { source = "dotfiles/ssh_config.tmpl", mode = "template" }

[history.reload]
"~/.config/hypr/**" = "hyprctl reload"
"~/.ssh/**" = "chmod 700 ~/.ssh"
```

```sh
$ mise dot apply --yes
mise files: created symlink ~/.config/hypr -> ~/.dotfiles/hypr
mise files: rendered ~/.dotfiles/ssh_config.tmpl to ~/.ssh/config
mise files: applied ~/.config/hypr, ~/.ssh/config
mise history: reload: hyprctl reload
mise history: reload: chmod 700 ~/.ssh
```

Details:

- The reload map is resolved from the trusted configuration layers before anything is written, as pull and rollback do, so nothing the apply writes can change which commands run.
- `mise dot apply --dry-run` writes nothing and runs no reload commands. An apply with nothing left to write runs none either.
- Entries are applied one at a time with no rollback of the earlier ones, so when the edit prompt is declined or a later entry fails (an unwritable parent, a full disk), the commands matching what was already written still run before the error is reported; the command still exits non-zero. A target is recorded for reload exactly when this apply changed its on-disk state: right after an existing target is cleared to make room (a write that then fails still leaves the old content gone); once an existing file that is overwritten in place — an edit target, or a file under a directory copy — has been opened for truncation (an open that fails on a read-only file or filesystem changes nothing and records nothing); and for a previously absent target, once it exists, including when the write that created it then failed. A target the failure never touched reloads nothing, and a directory copy that fails part-way reloads only for the files it did place.
- A symlink to a directory counts the files it exposes as written, so `"~/.config/hypr/**"` fires for a `"~/.config/hypr" = "dotfiles/hypr"` entry. That listing is best-effort: if the source cannot be walked after the link landed, mise warns and records the link itself.
- `mise dot sync` is unchanged: it never writes live files, so it has no reload to run.
- Internal: `system::files::apply`, `system::files::execute_apply` and `system::edits::apply` take a `written: &mut Vec<PathBuf>` the caller owns and append each target as it is mutated, so the list survives an error; `mise bootstrap`, `dot edit`, `dot add` and `system install` pass a throwaway vector and behave as before.

Validation: new e2e test `e2e/cli/test_dotfiles_apply_reload` (copy, directory symlink and edit entries each trigger their command once; an unmatched glob never runs; dry run and a no-op apply create no marker; two globs mapping to one command run it once; a second copy entry failing on an unwritable parent reloads the first entry only and the command fails; a directory copy failing on an unwritable subdirectory reloads only for the file it wrote). Unit tests for recording in `apply_one` (a partial directory copy, a write that fails before touching its target, an existing target cleared before a failed write, a failed write against an absent target, a read-only existing file that cannot be opened for an in-place overwrite or edit, a write that created its target and then failed, and a directory symlink whose source cannot be walked) and for the home-only path normalization. `cargo test --bin mise dotfiles`, `cargo test --bin mise system::files`, `cargo test --bin mise system::edits`, clippy with `-D warnings` and `mise run lint-fix` pass.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5-1; version: unavailable.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Runs user-defined shell commands and broadens dotfile apply/copy mutation tracking; behavior on partial failure and permission errors is nuanced but covered by tests.
> 
> **Overview**
> **`mise dot apply` now runs `[history.reload]` commands** for paths it actually changed, aligning apply with pull, rollback, and undo. Reload rules are loaded before any writes; **`--dry-run` and no-op applies run no reloads**. Matching commands run once after all writes, including when apply fails partway—only successfully mutated targets count.
> 
> Apply plumbing now **tracks written paths** through `system::files::apply` / `system::edits::apply` (bootstrap, `dot add`/`edit`, and shell install pass a discard buffer, so they still do not reload). Recording is tied to the first on-disk mutation (cleared targets, in-place truncations, new files). **Directory symlinks** also list exposed files so globs like `~/.config/hypr/**` match; listing failures are warned and fall back to the link path. **`reload_path`** normalizes `$HOME` for glob matching.
> 
> Directory **copy** paths were tightened so sources are opened and validated before replacing symlinks or truncating read-only targets, reducing cases where a failed copy leaves a destination missing or unchanged paths incorrectly marked as written.
> 
> Docs describe apply reload behavior; **e2e `test_dotfiles_apply_reload`** and unit tests cover dry-run, deduplication, partial failure, and recording edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 66bec34edb9be14ed3ee7050ad45c13d3ec25609. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **New Features**
  - `mise dot apply` now runs matching history reload commands after writing dotfiles, including copied, symlinked, templated, and edited targets.
  - Reload commands run only for files actually changed and execute once per matching command.

- **Bug Fixes**
  - Failed or partial applications now correctly reload files that were successfully written.
  - Dry runs and `mise dot sync` no longer trigger reload commands or modify live files.

- **Documentation**
  - Added guidance describing reload behavior across apply, sync, pull, rollback, and undo operations.

### CODE REVIEW

File: src/cli/dotfiles/apply.rs [hunk 1]
```diff
@@ -75,21 +80,30 @@ impl DotfilesApply {
                 force_hint: "use --force",
                 yes: self.yes,
             };
-            if !system::files::apply(&config, &files, &opts, &secrets)? {
-                return Ok(false);
-            }
+            let outcome = system::files::apply(&config, &files, &opts, &secrets)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/dotfiles/apply.rs]: Fixed in 433208dda. Pull and rollback also skip reload on a mid-operation failure (their `?`/`bail!` returns before `run_reload`), but apply has no journal to recover from, so `files::apply`/`edits::apply` now append each written target to a caller-owned `&mut Vec<PathBuf>` (listed before the write) and `mise dot apply` reloads whatever that holds before propagating the error; the e2e test covers a second copy entry failing on an unwritable parent while the first entry still reloads. Dry runs still reload nothing.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 2]
```diff
@@ -2351,13 +2361,18 @@ pub(crate) fn execute_apply(
                 );
             }
             let pending = journal::begin_changes_with(DOTFILES_PART, &item, paths)?;
+            written.push(link.target.clone());
             file::remove_file(&link.target)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:2364]: Fixed in 2a787910e: the pre-write `written_targets` listing is gone; `apply_one` now appends each path right after its own copy, link, render or removal succeeds (per file for directory copies and symlink-each, and edits once committed), so an entry that fails before touching its target reloads nothing and a partial directory copy reloads only for the files it placed. The e2e test now asserts the failed target's command does not run, plus a directory copy failing part-way; a unit test covers `apply_one` directly.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 3]
```diff
@@ -2351,13 +2361,18 @@ pub(crate) fn execute_apply(
...
             journal::commit_changes(pending);
         }
     }
     for (req, rendered) in &plan.todo {
         let pending =
             journal::begin_changes_with(DOTFILES_PART, &req.target_raw, touched_paths(req)?)?;
+        // listed before the write, so a write that fails part-way still
+        // reports its targets (and a symlink-each entry's stale links, which
+        // are gone once it converges)
+        written.extend(written_targets(req)?);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 2a787910e: `apply_one` records each target at its individual copy/link/removal (per file inside the directory-copy and symlink-each loops) and edits once committed, with one deliberate difference from the wording here: the path is appended right *after* the mutation succeeds rather than before it, since the documented behavior is to reload only for targets actually written and a failed mutation should not reload.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 4]
```diff
@@ -3406,6 +3424,14 @@ fn apply_one(req: &FileRequest, rendered: Option<&str>) -> Result<()> {
...
             remove_existing(&req.target)?;
             link_path(&req.source, &req.target, true)?;
+            written.push(req.target.clone());
+            if req.source.is_dir() {
+                written.extend(
+                    walk_source_files(req)?
+                        .into_iter()
+                        .map(|(_, target)| target),
+                );
+            }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:3439]: Fixed in 190ee5401: the post-link `walk_source_files` is best-effort now — on a walk error `apply_one` warns and keeps just the link path recorded, so the apply still succeeds and the journal commits; covered by a unit test with an unreadable subdirectory under the source.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 5]
```diff
@@ -3451,23 +3478,27 @@ fn apply_one(req: &FileRequest, rendered: Option<&str>) -> Result<()> {
                         file::remove_file(&target)?;
                     }
                     file::copy(&source, &target)?;
+                    written.push(target);
                 }
             } else {
                 remove_existing(&req.target)?;
                 file::copy(&req.source, &req.target)?;
+                written.push(req.target.clone());
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 190ee5401: each path is now recorded at its first mutation — `remove_existing` returns whether it cleared something and `replace_recorded` records the target right after that removal (or after the write when nothing was there), directory copies record an existing file before the in-place truncating copy, and edits record before `file::write` when the file exists (the journal preimage is only restored by a later `recover_pending` run, not inline, so after-commit would have missed a truncated file). Untouched targets are still never recorded; unit test extended with an existing target cleared before a failed write (recorded) and the same failure against an absent target (not recorded).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/edits.rs [hunk 6]
```diff
@@ -1130,19 +1138,31 @@ fn edit_paths(path: &Path) -> Vec<(PathBuf, Capture)> {
...
         file::read_to_string(&req.path)?
     } else {
         String::new()
     };
     let out = apply_to_string(req, desired, &text)?;
+    if existed {
+        written.push(req.path.clone());
+    }
     // file::write truncates in place, preserving the file's permissions
     file::write(&req.path, &out)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/edits.rs]: Fixed in 925dad0a3. Checked `src/file.rs`: `file::write` is `fs::write` and `file::copy` is `fs::copy` — both open the destination with truncation in place, no temp-file rename — so the first mutation of an existing file is the open, not the write. Edit targets and existing files under a directory copy are now opened explicitly with `OpenOptions::write(true).truncate(true)`, recorded once that open succeeds, then written (`write_all` / `io::copy` plus the source permission bits `fs::copy` set); a failed open on a read-only file records nothing. Unit tests added for a 0444 target in both the edit and dir-copy paths (skipped when running as root). The `remove_existing` rule is unchanged, and this closes the mutation-timing question from both sides: a path is recorded iff this apply changed its on-disk state.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 7]
```diff
@@ -3447,27 +3477,38 @@ fn apply_one(req: &FileRequest, rendered: Option<&str>) -> Result<()> {
...
                     if target.is_symlink() {
                         file::remove_file(&target)?;
                     }
+                    if existed {
+                        written.push(target.clone());
+                    }
                     file::copy(&source, &target)?;
+                    if !existed {
+                        written.push(target);
+                    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs:3510]: Fixed in 925dad0a3: the in-place overwrite of an existing edit target or directory-copy file is now an explicit open-for-truncation, then record, then write (`overwrite_recorded` in files.rs, the `existed` branch in edits.rs), so an open that fails on permissions records nothing; covered by 0444-target unit tests in both paths.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 8]
```diff
@@ -3516,11 +3558,33 @@ fn prune_stale_links(req: &FileRequest) -> Result<()> {
...
+    written: &mut Vec<PathBuf>,
+    write: impl FnOnce() -> Result<()>,
+) -> Result<()> {
+    let removed = remove_existing(target)?;
+    if removed {
+        written.push(target.to_path_buf());
+    }
+    write()?;
+    if !removed {
+        written.push(target.to_path_buf());
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 925dad0a3: a new `create_recorded` helper runs the write against a previously absent target and records it if it exists afterwards (`symlink_metadata` ok), whether the write succeeded or failed after creating the file; `replace_recorded`, the directory-copy loop and the edit write all go through it, completing the rule from both sides — recorded iff this apply changed the target's on-disk state. Unit test covers a closure that creates the file then errors (recorded), one that creates nothing (not recorded), and a plain success.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 9]
```diff
@@ -3516,15 +3554,76 @@ fn prune_stale_links(req: &FileRequest) -> Result<()> {
...
+    let mut to = std::fs::OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .open(target)
+        .wrap_err_with(failed)?;
+    written.push(target.to_path_buf());
+    std::io::copy(&mut from, &mut to).wrap_err_with(failed)?;
+    to.set_permissions(from.metadata()?.permissions())
+        .wrap_err_with(failed)?;
     Ok(())
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 61f4bde7f: `overwrite_recorded` now opens the source and requires `metadata().is_file()` (the same check `fs::copy` performs) before it opens the destination for truncation, so a directory symlink or other non-file source fails with the destination untouched and nothing recorded; unit test covers a dir-symlink source entry over an existing regular file. The edit path already `read_to_string`s the target before opening it for writing, so a non-file there fails first as well.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: e2e/cli/test_dotfiles_apply_reload [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+assert "cat ~/reloads" "app"
+
+# a single command runs once even when several written targets and globs
+# match it
+rm ~/reloads
+echo 'theme = dark' >"$MISE_CONFIG_DIR/dotfiles/app/settings.toml"
+cat <<EOF >"$config"
+[dotfiles]
+"~/.config/app/settings.toml" = { source = "dotfiles/app/settings.toml", mode = "copy" }
+"~/.config/app/keys.toml" = { content = "leader = space\n" }
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e/cli/test_dotfiles_apply_reload:61]: Done in 29bc7bccd: the CLAUDE.md rule is about end-of-test cleanup (the harness handles that), and these `rm`s were resets between cases rather than cleanup — but each case now appends to its own marker (`~/reloads-dry-run`, `-apply`, `-noop`, `-changed`, `-once`, `-partial`, `-multi`) so no reset is needed at all; every assertion is equivalent, and the two `chmod 755` restores went too since the harness does not need them.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 11]
```diff
@@ -3448,26 +3478,33 @@ fn apply_one(req: &FileRequest, rendered: Option<&str>) -> Result<()> {
...
                     }
                     if target.is_symlink() {
+                        // a link is replaced: recorded once it is gone
                         file::remove_file(&target)?;
+                        written.push(target.clone());
+                        file::copy(&source, &target)?;
+                    } else if target.is_file() {
+                        overwrite_recorded(&source, &target, written)?;
+                    } else {
+                        create_recorded(&target, written, || file::copy(&source, &target))?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in ce3a623ad — this was a real case: a pre-existing directory at the target reached the create path, `fs::copy` failed without touching it, and the after-the-fact existence check recorded it anyway. The branch now checks `symlink_metadata` first: `NotFound` goes through `create_recorded`, an existing non-file target is recorded only after `file::copy` succeeds, and other metadata errors propagate; `overwrite_recorded` still handles regular files. Unit test covers a directory at the target where the copy wants a file (apply errs, directory intact, nothing recorded).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/files.rs [hunk 12]
```diff
@@ -3448,26 +3478,50 @@ fn apply_one(req: &FileRequest, rendered: Option<&str>) -> Result<()> {
                         file::create_dir_all(parent)?;
                     }
                     if target.is_symlink() {
+                        // a link is replaced: recorded once it is gone. The
+                        // source is checked first, as `fs::copy` would, so
+                        // one that cannot be copied leaves the link alone
+                        ensure_copy_source(&source, &target)?;
                         file::remove_file(&target)?;
+                        written.push(target.clone());
+                        file::copy(&source, &target)?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/files.rs]: Fixed in 66bec34ed: `ensure_copy_source` is replaced by a `CopySource` handle that `File::open`s the source and checks `metadata().is_file()` on that handle before the destination is touched; the link branch removes the link, records it, `File::create`s the target and `io::copy`s from the already-open handle (permissions from the handle metadata), and `overwrite_recorded` uses the same helper, so nothing in the loop reopens the source. The two non-file-source unit tests now exercise this path and stay green.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
